actors/miner/src/lib.rs

```diff
@@ -4302,48 +4207,140 @@ fn verify_windowed_post(
-        params.interactive_epoch,
-        &entropy,
-    )?;
+impl Into<AggregateSealVerifyInfo> for SealVerifyInfoInput {
+    fn into(self) -> AggregateSealVerifyInfo {
+        AggregateSealVerifyInfo {
+            sector_number: self.sector_number,
+            randomness: self.randomness,
+            interactive_randomness: self.interactive_randomness,
+            sealed_cid: self.sealed_cid,
+            unsealed_cid: self.unsealed_cid,
+        }
+    }
+}
 
-    let commd = unsealed_cid.get_cid(params.registered_seal_proof)?;
+fn get_valid_verify_infos(
+    rt: &impl Runtime,
+    precommits: &[SectorPreCommitOnChainInfo],
+    proof_lengths: &[Option<usize>],
+) -> Result<(BatchReturn, Vec<SealVerifyInfoInput>), ActorError> {
+    let mut batch = BatchReturnGen::new(precommits.len());
 
-    Ok(SealVerifyInfo {
-        registered_proof: params.registered_seal_proof,
-        sector_id: SectorID { miner: miner_actor_id, number: params.sector_num },
-        deal_ids: params.deal_ids,
-        interactive_randomness: Randomness(interactive_randomness.into()),
-        proof: params.proof,
-        randomness: Randomness(randomness.into()),
-        sealed_cid: params.sealed_cid,
-        unsealed_cid: commd,
-    })
+    let mut verify_infos = vec![];
+    for (i, precommit) in precommits.iter().enumerate() {
+        let msd =
+            max_prove_commit_duration(rt.policy(), precommit.info.seal_proof).ok_or_else(|| {
+                actor_error!(
+                    illegal_state,
+                    "no max seal duration for proof type: {}",
+                    i64::from(precommit.info.seal_proof)
+                )
+            })?;
+        let prove_commit_due = precommit.pre_commit_epoch + msd;
+        if rt.curr_epoch() > prove_commit_due {
+            log::warn!(
+                "skipping commitment for sector {}, too late at {}, due {}",
+                precommit.info.sector_number,
+                rt.curr_epoch(),
+                prove_commit_due,
+            );
+            batch.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
+        } else {
+            batch.add_success();
+        }
+        // All seal proof types should match
+        // XXX this is an aggregate specific check, ok to include in general case ?
+        if i >= 1 {
+            let prev_seal_proof = precommits[i - 1].info.seal_proof;
+            if prev_seal_proof != precommit.info.seal_proof {
+                return Err(actor_error!(
+                    // XXX is this really illegal state and not illegal argument?
+                    illegal_state,
+                    "seal proof group for verification contains mismatched seal proofs {} and {}",
+                    i64::from(prev_seal_proof),
+                    i64::from(precommit.info.seal_proof)
+                ));
+            }
+        }
+        // XXX this is a non-aggregate specific check, ok to include in general case?
```

Comment:
It's a bit out of place here, indicated by the extra optional parameter that's data to be checked, rather than a flag to switch behaviour. I'd probably pull this one out, but not blocking on it.